Better spa override

Seanny123 opened the pull request:
Resolves part of #559

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -53,11 +69,12 @@ def gather_data(self, t, x):
-                v = np.dot(self.vocab_out.transform_to(self.vocab_in), v)
+            if self.loop_in:
+                if self.vocab_in is not self.vocab_out:
+                    v = np.dot(self.vocab_out.transform_to(self.vocab_in), v)
```
That call to ```transform_to``` is going to be incredibly expensive to compute every timestep.  If we do something like this, it'd be a huge speed improvement to do it outside the loop, I think.

I'm also unsure that we even need this complexity.  The one case where it could be used is an AssociativeMemory with a different vocab for the input and output, and I'd be tempted to just remove that case from the whitelist.

tcstewar replied:
Although before I suggest removing it from the whitelist, how well did it work when you tried it?  What sort of things did you try?

Seanny123 replied:
I tried it with a normal clean-up memory and it worked fine, but today I realised that it didn't work with an associative memory where the input_keys don't match the output_keys.

tcstewar replied:
Ah, good point.  Yup, I'd remove that from the whitelist.

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,34 @@ def __init__(self, obj, **kwargs):
         # a value of 'None' means do not override
         self.override_target = None
-        self.vocab_in = obj.inputs[self.target][1]
+        loop_in_whitelist = [spa.Buffer, spa.Memory, spa.State, spa.AssociativeMemory]
```
All of these should only be whitelisted for the case when ```self.target == 'default'```

Seanny123 replied:
That's how `self.target` is defined in `SpaPlot`.

tcstewar replied:
> That's how self.target is defined in SpaPlot.

No it's not.  The line in SpaPlot says ```self.target = kwargs.get('args', 'default')``` which means that it *defaults* to ```'default'```, but if someone passes it in to the constructor something like ```Pointer(target='A')``` then it'll display that output instead.  Admittedly, we don't currently use that argument, but the fact that it's there and we have specific code in the constructor to handle setting it means that we shouldn't assume that it's always set to ```'default'```.

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,34 @@ def __init__(self, obj, **kwargs):
+        if type(obj) in loop_in_whitelist:
+            self.vocab_in = obj.inputs[self.target][1]
+            self.loop_in = True
+            self.default_override_val = self.vocab_in.parse('0').v
```
since we have ```size_out``` here, this would be clearer as ```np.zeros(self.size_out)```

Seanny123 commented:
Ready for review

Seanny123 commented:
Hey Terry, don't know if you missed my last ping, but this is the last bit that needs review before we can make a Nengo GUI release.

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,35 @@ def __init__(self, obj, **kwargs):
         # a value of 'None' means do not override
         self.override_target = None
-        self.vocab_in = obj.inputs[self.target][1]
+        loop_in_whitelist = [spa.Buffer, spa.Memory, spa.State]
```
There should be a comment indicating what this whitelist is, and what conditions there are if someone were to add something new onto the whitelist (and also why we have this whitelist in the first place)

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,35 @@ def __init__(self, obj, **kwargs):
+
+        self.vocab_in = None
+        if type(obj) in loop_in_whitelist and self.target == 'default':
+            self.vocab_in = obj.inputs[self.target][1]
```
Given the whitelist, ```self.vocab_in``` is always identical to ```self.vocab_out```, so this variable isn't needed.

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,35 @@ def __init__(self, obj, **kwargs):
+        if type(obj) in loop_in_whitelist and self.target == 'default':
+            self.vocab_in = obj.inputs[self.target][1]
+            self.loop_in = True
+            self.size_out = self.vocab_in.dimensions
```
Due to the comment above, we don't need ```size_out``` either, as we can just always use ```self.vocab_out.dimensions```.

tcstewar commented:
Just to check, what was your reasoning for removing AssociativeMemory from the whitelist?  I'm not sure I disagree with that approach, but I'm just making sure you didn't accidentally do it...  :)

Seanny123 commented:
I removed it on purpose. The approach of looping-in when there was a
different vocab on the input and ouput, even if the connection between them
tried to transform between the two vocabs, didn't work, so I removed it
from the whitelist.

On Mon, Oct 26, 2015 at 3:56 PM, tcstewar <notifications@github.com> wrote:

> Just to check, what was your reasoning for removing AssociativeMemory from
> the whitelist? I'm not sure I disagree with that approach, but I'm just
> making sure you didn't accidentally do it... :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nengo/nengo_gui/pull/561#issuecomment-151265220>.
>

tcstewar commented:
> I removed it on purpose. The approach of looping-in when there was a
> different vocab on the input and ouput, even if the connection between them
> tried to transform between the two vocabs, didn't work, so I removed it
> from the whitelist.

Fair enough.  I assume that also means that you felt that checking for the (common) use-case where the input vocab and output vocab are the same wasn't worth it?

Seanny123 commented:
Exactly.

On Mon, Oct 26, 2015 at 4:05 PM, tcstewar <notifications@github.com> wrote:

> I removed it on purpose. The approach of looping-in when there was a
> different vocab on the input and ouput, even if the connection between them
> tried to transform between the two vocabs, didn't work, so I removed it
> from the whitelist.
>
> Fair enough. I assume that also means that you felt that checking for the
> (common) use-case where the input vocab and output vocab are the same
> wasn't worth it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nengo/nengo_gui/pull/561#issuecomment-151267328>.
>

Seanny123 commented:
Fixed all your comments. If you want to explore the network for yourself, this is what I tested with.
```
import nengo
from nengo import spa

D = 32
vo = spa.Vocabulary(D)
vo.parse("A + B + C")
model = spa.SPA(vocabs=[vo])


with model:
    model.in_A = spa.State(D)
    model.in_B = spa.State(D)
    model.bind = spa.Bind(D, invert_b=True)
    model.ass = spa.AssociativeMemory(vo)
    model.out = spa.State(D)
    model.cortical = spa.Actions(
        'bind_A = in_A',
        'bind_B = in_B',
        'ass = bind',
        'out = ass'
    )
    model.c_act = spa.Cortical(model.cortical)
```

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,31 @@ def __init__(self, obj, **kwargs):
+        # All other networks have their value inserted on the output
+        self.loop_in_whitelist = [spa.Buffer, spa.Memory, spa.State]
+
+        self.obj = obj
```
this is already being set in the parent class

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,31 @@ def __init__(self, obj, **kwargs):
+        # The white list indicates the networks whose user-defined
+        # over-ride value can be inserted on the input
+        # thus the value loops in from the output to the input.
+        # All other networks have their value inserted on the output
```
Add in a mention that the whole point of looping in on the input is so that the actual neural activity of this module will be changed by setting its value, rather than just the output of the module.

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -16,17 +15,31 @@ def __init__(self, obj, **kwargs):
+                                   size_out=self.vocab_out.dimensions)
             self.conn1 = nengo.Connection(output, self.node, synapse=0.01)
-            self.conn2 = nengo.Connection(self.node, input, synapse=0.01)
+            if(type(self.obj) in self.loop_in_whitelist
```
spacing

Seanny123 commented:
Fixed most recent comments

tcstewar commented on nengo_gui/components/pointer.py:
```diff
@@ -1,32 +1,48 @@
             self.conn1 = nengo.Connection(output, self.node, synapse=0.01)
-            self.conn2 = nengo.Connection(self.node, input, synapse=0.01)
+            if (type(self.obj) in self.loop_in_whitelist
+                and self.target == 'default'):
```
This line does not distinguish itself from the indented line below, making it very hard to parse.

Note that this is exactly the one annoying case in Python indenting (an if statement with a parenthesis continuation, meaning that the natural amount you'd want to indent the second part of the first line is exactly four spaces).  I normally just rewrite the logic a bit to avoid having to do the line continuation.

Seanny123 replied:
Now that you've reminded me, I do recall you mentioning this before. I've
switched the logic and will remember this from now on.

On Mon, Oct 26, 2015 at 5:11 PM, tcstewar <notifications@github.com> wrote:

> In nengo_gui/components/pointer.py
> <https://github.com/nengo/nengo_gui/pull/561#discussion_r43053621>:
>
> >              self.conn1 = nengo.Connection(output, self.node, synapse=0.01)
> > -            self.conn2 = nengo.Connection(self.node, input, synapse=0.01)
> > +            if (type(self.obj) in self.loop_in_whitelist
> > +                and self.target == 'default'):
>
> This line does not distinguish itself from the indented line below, making
> it very hard to parse.
>
> Note that this is exactly the one annoying case in Python indenting (an if
> statement with a parenthesis continuation, meaning that the natural amount
> you'd want to indent the second part of the first line is exactly four
> spaces). I normally just rewrite the logic a bit to avoid having to do the
> line continuation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nengo/nengo_gui/pull/561/files#r43053621>.
>

tcstewar commented:
Looks good to me!
